contracts/modules/staking/StakingPool.sol

```diff
@@ -1223,16 +1216,54 @@ contract StakingPool is IStakingPool, ERC721 {
     return ownerOf(0);
   }
 
-  /* management */
+  /* pool management */
+
+  function setProducts(ProductParams[] memory params) external onlyManager {
+    _setProducts(params);
+  }
+
+  function _setProducts(ProductParams[] memory params) internal {
+    uint32 _targetWeight = targetWeight;
+    for (uint i = 0; i < params.length; i++) {
+      ProductParams memory _param = params[i];
+      StakedProduct memory _product = products[_param.productId];
+      if (_product.nextPriceUpdateTime == 0) {
+        Product memory coverProduct = ICover(coverContract).products(_param.productId);
+        require(coverProduct.initialPriceRatio > 0, "Failed to get initial price for product");
+        _product.nextPrice = coverProduct.initialPriceRatio;
+        _product.nextPriceUpdateTime = uint32(block.timestamp);
+      }
+
+      if (_param.setPrice) {
+        require(_param.targetPrice <= SURGE_PRICE_DENOMINATOR, "Target price too high");
+        _product.targetPrice = _param.targetPrice;
+      }
+
+      if (_param.setWeight) {
+          require(_param.targetWeight <= WEIGHT_DENOMINATOR, "Cannot set weight beyond 1");
 
-  function addProducts(ProductParams[] memory params) external onlyManager {
-    totalSupply = totalSupply;  // To silence view fn warning. Remove once implemented
-    params;
+          if (_product.targetWeight < _param.targetWeight) {
+             _targetWeight += _param.targetWeight - _product.targetWeight;
+          } else {
+            _targetWeight -= _product.targetWeight - _param.targetWeight;
+          }
+          _product.targetWeight = _param.targetWeight;
+      }
+      products[_param.productId] = _product;
+    // End for loop
+    }
+    require(_targetWeight <= 2000, "Target weight above 20");
+    targetWeight = _targetWeight;
   }
 
-  function removeProducts(uint[] memory productIds) external onlyManager {
-    totalSupply = totalSupply;  // To silence view fn warning. Remove once implemented
-    productIds;
+  function _setInitialPrices(ProductInitializationParams[] memory params) internal returns (ProductParams[] memory res) {
```

Comment:
params is copied over once here when it gets passed (it's potentially large), I don't see a problem with inlining this loop since it's simple enough

Reply:
@shark0der  did a quick experiment.

Apparently the array is passed by reference so it's not copied. This is fine then.

![image](https://user-images.githubusercontent.com/1534891/194304991-b4854cd2-b22d-44a2-9c0f-405f39a85f25.png)
![image](https://user-images.githubusercontent.com/1534891/194305013-5b04b823-6492-4baa-a667-a0b1875a35e9.png)